### PR TITLE
Channel<T> improvements

### PR DESCRIPTION
**Pull request by sixlettervariables**
This PR adds two improvements to `Channel<T>` usage within csnats:

1. Batched `get` operation.
2. PING/PONG use a specially designed channel for PONG relies (`SingleUseChannel<T>`).

A `tryAdd` member has been added to `Channel<T>` as well, to be used in a subsequent perf-related PR.

Using the supplied Benchmark utility, best-of-three runs were done comparing against master (read: not entirely scientific). My specs are the following: Processor=Intel(R) Xeon(R) CPU E5-2620 v3 2.40GHzIntel(R) Xeon(R) CPU E5-2620 v3 2.40GHz, ProcessorCount=24

The best of each results are given below:
```
                       | Baseline            | Channel<T>       | Delta           | Speedup  
----------- | -------- | ---------- | ------ | ------- | ------ | ------- | ----- | ----------- | ------------
            | Count    | msgs/s     | kb/s   | msgs/s  | kb/s   | msgs/s  | kb/s  | msgs/s      | kb/s        
PubOnlyNo   | 10000000 | 2302097    | 0      | 2549885 | 0      | 247788  | 0     | 1.107635777 |             
PubOnly8b   | 10000000 | 1982378    | 15487  | 2009341 | 15697  | 26963   | 210   | 1.013601341 | 1.01355976  
PubOnly32b  | 10000000 | 1664874    | 52027  | 1774239 | 55444  | 109365  | 3417  | 1.065689656 | 1.065677437 
PubOnly256b | 10000000 | 730606     | 182651 | 881915  | 220478 | 151309  | 37827 | 1.207100681 | 1.207099879 
PubOnly512b | 10000000 | 435323     | 217661 | 570142  | 285071 | 134819  | 67410 | 1.309698775 | 1.309701784 
PubOnly1k   | 1000000  | 233951     | 233951 | 304485  | 304485 | 70534   | 70534 | 1.301490483 | 1.301490483 
PubOnly4k   | 500000   | 57334      | 229336 | 79176   | 316704 | 21842   | 87368 | 1.380960687 | 1.380960687 
PubOnly8k   | 100000   | 28266      | 226128 | 37534   | 300272 | 9268    | 74144 | 1.327885092 | 1.327885092 
PubSubNo    | 10000000 | 603724     | 0      | 810147  | 0      | 206423  | 0     | 1.341916174 |             
PubSub8b    | 10000000 | 627404     | 4901   | 836156  | 6532   | 208752  | 1631  | 1.332723413 | 1.332789227 
PubSub32b   | 10000000 | 680405     | 21262  | 770157  | 24067  | 89752   | 2805  | 1.131909671 | 1.131925501 
PubSub256b  | 10000000 | 386603     | 96650  | 466495  | 116623 | 79892   | 19973 | 1.206651268 | 1.206652871 
PubSub512b  | 500000   | 235251     | 117625 | 270597  | 135298 | 35346   | 17673 | 1.150248033 | 1.150248672 
PubSub1k    | 500000   | 112693     | 112693 | 135642  | 135642 | 22949   | 22949 | 1.203641752 | 1.203641752 
PubSub4k    | 500000   | 28192      | 112768 | 35563   | 142252 | 7371    | 29484 | 1.261457151 | 1.261457151 
PubSub8k    | 100000   | 15635      | 125080 | 18280   | 146240 | 2645    | 21160 | 1.16917173  | 1.16917173  
ReqReplNo   | 20000    | 4447       | 0      | 6130    | 0      | 1683    | 0     | 1.378457387 |             
ReqRepl8b   | 10000    | 4080       | 31     | 5654    | 44     | 1574    | 13    | 1.385784314 | 1.419354839 
ReqRepl32b  | 10000    | 3905       | 122    | 5308    | 165    | 1403    | 43    | 1.359282971 | 1.352459016 
ReqRepl256b | 5000     | 3912       | 978    | 4750    | 1187   | 838     | 209   | 1.214212679 | 1.213701431 
ReqRepl512b | 5000     | 3396       | 1698   | 4438    | 2219   | 1042    | 521   | 1.306831567 | 1.306831567 
ReqRepl1k   | 5000     | 3767       | 3767   | 5817    | 5817   | 2050    | 2050  | 1.544199628 | 1.544199628 
ReqRepl4k   | 5000     | 2483       | 9932   | 4316    | 17264  | 1833    | 7332  | 1.738219895 | 1.738219895 
ReqRepl8k   | 5000     | 2106       | 16848  | 3362    | 26896  | 1256    | 10048 | 1.596391263 | 1.596391263
```
Basically an average of 1.29x improvement on my machine with a batch size of 64. This was empirically found to be the best batch size on my machine. That may depend heavily on the workload.

### CODE REVIEW

File: NATS.Client/Channel.cs [hunk 1]
```diff
@@ -1,11 +1,72 @@
...
+            actualValue = value;
+            hasValue = true;
+            e.Set();
+        }
+
+        internal void reset()
+        {
+            hasValue = false;
+            e.Reset();
+            actualValue = default(T);
```
Comment (ColinSullivan1) [inline NATS.Client/Channel.cs]: Should `actualValue` be assigned before the Reset() ?  I'm thinking through the thread safety of `actualValue`...

File: NATS.Client/Channel.cs [hunk 2]
```diff
@@ -1,11 +1,72 @@
...
 using System;
+using System.Collections.Concurrent;
 using System.Collections.Generic;
 using System.Threading;
 
 namespace NATS.Client
 {
+    // Provides a Channel<T> implementation that only allows a single call to 'get'
+    // Used for ping-pong
+    internal sealed class SingleUseChannel<T>
```
Comment (ColinSullivan1) [inline NATS.Client/Channel.cs]: Can you add a comment indicating that this class is not entirely thread-safe and the caller must ensure thread safety in it's usage?  That being said, as is, this is ideal because it should be OK with the calling code, and is certainly performance oriented.

File: NATS.Client/Channel.cs [hunk 3]
```diff
@@ -16,101 +77,204 @@ internal sealed class Channel<T>
         readonly Object   qLock = new Object();
 
         bool   finished = false;
-        
+
+        public string Name { get; set; }
+
         internal Channel()
         {
+            Name = "Unnamed " + this.GetHashCode();
```
Comment (ColinSullivan1) [inline NATS.Client/Channel.cs]: Nitpick:  How about "Channel"?

File: NATS.Client/Options.cs [hunk 4]
```diff
@@ -366,6 +370,30 @@ public int SubscriberDeliveryTaskCount
...
+        /// Gets or sets the batch size for calling subscription handlers.
+        /// </summary>
+        /// <remarks>
+        /// When delivering messages to the subscriber, the batch size determines
+        /// how many messages could be retrieved from the internal subscription
+        /// queue at one time. This can allow higher performance from a single
+        /// subscriber by avoiding the locking overhead of one-at-a-time
+        /// retrieval from the queue.
+        /// </remarks>
+        public int SubscriptionBatchSize
```
Comment (ColinSullivan1) [inline NATS.Client/Options.cs:390]: Can you add this option in `ToString()`?

File: NATS.Client/Channel.cs [hunk 5]
```diff
@@ -1,11 +1,72 @@
...
+        public static void Return(SingleUseChannel<T> ch)
+        {
+            ch.reset();
+            if (Channels.Count < 1024) Channels.Add(ch);
+        }
+
+        bool hasValue = false;
+        T actualValue;
+        readonly ManualResetEventSlim e = new ManualResetEventSlim();
+
```
Comment (ColinSullivan1) [inline NATS.Client/Channel.cs]: In most cases I prefer these closer to the class definition, above.

File: NATS.Client/Channel.cs [hunk 6]
```diff
@@ -1,11 +1,72 @@
...
+            return new SingleUseChannel<T>();
+        }
+
+        public static void Return(SingleUseChannel<T> ch)
+        {
+            ch.reset();
+            if (Channels.Count < 1024) Channels.Add(ch);
+        }
+
+        bool hasValue = false;
```
Comment (ColinSullivan1) [inline NATS.Client/Channel.cs]: volatile?

**Review by ColinSullivan1 — COMMENTED** [review]
@sixlettervariables , Great stuff!  I had a few questions. 
 This is very critical code to the NATS client; I'll run your PR in some stress tests overnight.

File: NATS.Client/Channel.cs (hunk 3 above)
Comment (sixlettervariables) [reply NATS.Client/Channel.cs]: This made more sense when I named the channels! I've added that back.

File: NATS.Client/Options.cs (hunk 4 above)
Comment (sixlettervariables) [reply NATS.Client/Options.cs:390]: Done.

**Review by ColinSullivan1 — APPROVED** [review]
LGTM.  Excellent contribution - Thank you!
